### PR TITLE
Fix silu gradient vanishing at large-negative inputs

### LINKED ISSUE #39798: jax.nn.silu gradient returns zero for large negative float64 input
### Description

`jax.nn.silu` appears to return an incorrect zero gradient for a large negative finite `float64` input.

For SiLU / Swish,

```
silu(x) = x * sigmoid(x)
```

Its derivative is:

```
silu'(x) = sigmoid(x) + x * sigmoid(x) * (1 - sigmoid(x))
```

At `x = -709.0`, the true derivative is approximately `-8.614807714413836e-306`, which is finite and representable in `float64`. However, JAX returns `0.0`.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import math

os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("JAX_PLATFORMS", "cpu")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

try:
    import jax
    import jax.numpy as jnp
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

jax.config.update("jax_enable_x64", True)

x = -709.0
expected = -8.614807714413836e-306

out = float(jax.grad(lambda z: jax.nn.silu(z))(jnp.asarray(x, dtype=jnp.float64)))

print("jax:", out)
print("expected:", expected)

if not (math.isfinite(out) and abs(out - expected) <= 1e-6 * abs(expected)):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
jax: 0.0
expected: -8.614807714413836e-306
```

### Expected result

```
-8.614807714413836e-306
```

The expected derivative is finite and representable in `float64`. JAX instead returns `0.0`, suggesting that the backward pass flushes or loses the tiny but nonzero derivative for large negative inputs.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

### PR DESCRIPTION
**Pull request by Phoebus-Liu**
Fixes #39798

### Description

`jax.nn.silu` returned exactly `0.0` for its gradient at large-negative `float64` inputs, e.g. `x = -709.0`, where the true derivative is `-8.614807714413836e-306` — finite and representable.

```python
jax.config.update("jax_enable_x64", True)
jax.grad(jax.nn.silu)(jnp.asarray(-709.0, dtype=jnp.float64))
# before: 0.0
# after:  -8.614807714413836e-306
```

### Root cause

`silu(x) = x * sigmoid(x)` had no explicit derivative rule, so the gradient came from differentiating that implementation. Substituting `e = exp(x)` on the negative branch, `s = e / (1 + e)` and `1 - s = 1 / (1 + e)`, gives

```
silu'(x) = s * (1 + x * (1 - s)) = e * (1 + e + x) / (1 + e)^2
```

The product `e * (1 + e + x)` is the problem: `exp(-709)` is already subnormal, so the product underflows — and is flushed to zero entirely on backends that do not support subnormals.

### Fix

Give `silu` a `custom_jvp` rule that evaluates the derivative directly instead of relying on the differentiated primal:

* For `x >= 0`, `s = 1 / (1 + exp(-x))` and the textbook form `s * (1 + x * (1 - s))` is already accurate.
* For `x < 0`, the form `e * (1 + e + x) / (1 + e)^2` is used while `e` is still a normal number. Once `e` drops below the smallest normal, the product is instead built in log space as `sign(a) * exp(x + log|a|) / (1 + e)^2`, where `a = 1 + e + x`, so that no intermediate is subnormal.

Keeping the direct form wherever it is accurate means precision is unchanged (~1 ulp) across the well-conditioned range; only the previously-broken tail changes. In the tail the result is now accurate to ~5e-14 relative (dominated by the `exp`/`log` round trip) instead of being zero.

Complex inputs take a holomorphic branch that reproduces the previous behaviour exactly.

### Notes

* This fixes the *gradient* only. The primal `silu(-709.0)` still evaluates to `-0.0` because `sigmoid(-709.0)` underflows; that is a separate range limitation and is left unchanged to keep this fix minimal.
* `testSiluGradLargeNegative` is added to `tests/nn_test.py` with a strict tolerance (`rtol=1e-12, atol=0.0`); it returns exactly `0.0` without this change.

### Verification

* New regression test fails before / passes after (verified by temporarily removing the rule).
* `check_grads(jax.nn.silu, (-3.,), order=3, modes=['fwd','rev'])` passes, as do higher-order `grad`, `hessian`, `vmap` and `jit`.
* Complex gradient matches the analytic holomorphic derivative exactly.
* Dense sweep against a 60-digit `decimal` reference: max relative error ~1 ulp for `x > -708`, 4.9e-14 at `x = -709`.
* 48 activation/gradient tests in `tests/nn_test.py` pass. (The full file cannot be run in this environment: it aborts with a pre-existing `MemoryError`/segfault in the GPU-oriented `DotProductAttention`/`ScaledMatmul` tests, identical on an unmodified checkout.)

### CODE REVIEW

**Comment by jakevdp** [comment]
Thanks for looking at this – I don't think we're going to want to make this change. The following is pasted from a Gemini agent review of this PR; I think the issues it identifies are real, and the first is a particularly important blocker:

### Important Issues with the Current Implementation
1. **Severe backward-pass performance regression on a hot-path activation (`silu` / `swish`):**
   - Currently, `silu(x) = x * sigmoid(x)` lowers to `x * lax.logistic(x)`. In the backward pass, JAX reuses the primal `sigmoid(x)` output from the forward pass and computes the derivative in **3 basic arithmetic ops (`sub`, `mul`, `add`) with zero transcendental operations**.
   - With this `@custom_jvp`, the forward `sigmoid(x)` output (`ans`) is discarded, and every backward pass of `silu` unconditionally evaluates **both** branches of `jnp.where`, executing **4 transcendental calls (`3x jnp.exp`, `1x jnp.log`), 3 divisions, and 3 `jnp.where` selects**.
   - Because `jax.nn.silu` is used in the MLP block (SwiGLU) of nearly every modern LLM, adding 4 transcendental ops to every `silu` backward pass is a significant performance and memory/register regression to work around CPU subnormal flushing in a narrow 6-unit window (`x ∈ [-714, -708]` in `float64`). See also the policy discussion in #38651 and #37766 regarding XLA's flush-to-zero semantics for subnormals.
2. **Introduces `NaN` gradients at `±inf`:**
   - **At `x = +inf`**: In the `x >= 0` branch, `e_pos = exp(-inf) = 0.0` and `s_pos = 1.0`, so `grad_pos = s_pos * (1.0 + x_arr * (1.0 - s_pos))` evaluates `1.0 + (+inf * 0.0) = NaN` (whereas `jax.nn.silu` should have derivative `1.0` as `x -> +inf`).
   - **At `x = -inf`**: In the `x < 0` log-space branch, `x_neg = -inf` and `a = 1.0 + 0.0 + (-inf) = -inf`, so `jnp.exp(x_neg + jnp.log(jnp.abs(a)))` evaluates `exp(-inf + inf) = NaN` (whereas `silu'(-inf)` should be `0.0`).
3. **Breaks higher-order autodiff (`jax.grad(jax.grad(jax.nn.silu))`) via the `jnp.where` trap:**
   - In `_silu_grad`, `jnp.where(e_neg >= tiny, e_neg * a / denom, jnp.sign(a) * jnp.exp(x_neg + jnp.log(jnp.abs(a))) / denom)` does not use the double-`where` pattern to sanitize `a` before passing it into `jnp.log(jnp.abs(a))`.
   - Even when `e_neg >= tiny` is `True`, reverse-mode AD differentiates both branches of `jnp.where`. At the root of `a = 1 + exp(x) + x = 0` (`x ≈ -1.2784645`), the derivative of `jnp.log(jnp.abs(a))` (`1 / a`) in the unselected branch evaluates to `inf`, causing `0.0 * inf = NaN` in `jax.grad(jax.grad(jax.nn.silu))`.

**Comment by Phoebus-Liu** [comment]
Hi @jakevdp,

Thanks for the detailed write-up — that's a fair call, and I accept it. Avoiding a backward-pass regression on a hot-path activation like `silu`/`swish` clearly outweighs the numerical edge case this PR targeted, and trading three arithmetic ops for a transcendental in the VJP is not a good trade.

I'm closing this out; no further action needed on your side. Appreciate you taking the time to review it and to explain the reasoning so clearly.
